### PR TITLE
Added a `neo4j-admin memrec` command, which can print recommendations for the Neo4j heap and pagecache memory settings.

### PR DESCRIPTION
**Pull request by chrisvest**
This is an initial stab at having automated recommendations for the Neo4j memory settings.

It looks like this:

```
$ bin/neo4j-admin help
usage: neo4j-admin <command>

Manage your Neo4j instance.

environment variables:
    NEO4J_CONF    Path to directory which contains neo4j.conf.
    NEO4J_DEBUG   Set to anything to enable debug output.
    NEO4J_HOME    Neo4j home directory.
    HEAP_SIZE     Set size of JVM heap during command execution.
                  Takes a number and a unit, for example 512m.

available commands:

General
    check-consistency
        Check the consistency of a database.
    help
        This help text, or help for the command specified in <command>.
    import
        Import from a collection of CSV files or a pre-3.0 database.
    memrec
        Print Neo4j heap and pagecache memory settings recommendations.
    store-info
        Prints information about a Neo4j database store.

Authentication
    set-default-admin
        Sets the default admin user when no roles are present.
    set-initial-password
        Sets the initial password of the initial admin user ('neo4j').

Offline backup
    dump
        Dump a database into a single-file archive.
    load
        Load a database from an archive created with the dump command.

Use neo4j-admin help <command> for more details.
```

And

```
$ bin/neo4j-admin help memrec
usage: neo4j-admin memrec [--memory=<16g>]

environment variables:
    NEO4J_CONF    Path to directory which contains neo4j.conf.
    NEO4J_DEBUG   Set to anything to enable debug output.
    NEO4J_HOME    Neo4j home directory.
    HEAP_SIZE     Set size of JVM heap during command execution.
                  Takes a number and a unit, for example 512m.

Print heuristic memory setting recommendations for the Neo4j JVM heap and
pagecache. The heuristic is based on the total memory of the system the command
is running on, or on the amount of memory specified with the --memory argument.
The heuristic assumes that the system is dedicated to running Neo4j. If this is
not the case, then use the --memory argument to specify how much memory can be
expected to be dedicated to Neo4j.

The output is formatted such that it can be copy-posted into the neo4j.conf
file.

options:
  --memory=<16g>   Recommend memory settings with respect to the given amount of
                  memory, instead of the total memory of the system running the
                  command. [default:16g]
```

And

```
$ bin/neo4j-admin memrec --memory 12g
# Memory settings recommendation from neo4j-admin memrec:
#
# Assuming the system is dedicated to running Neo4j and has 12g of memory,
# we recommend a heap size of around 4600m, and a page cache of around 4g,
# and that about 3600m is left for the operating system, and the native memory
# needed by Lucene and Netty.
#
# Tip: If the indexing storage use is high, e.g. there are many indexes or most
# data indexed, then it might advantageous to leave more memory for the
# operating system.
#
# Tip: The more concurrent transactions your workload has and the more updates
# they do, the more heap memory you will need. However, don't allocate more
# than 31g of heap, since this will disable pointer compression, also known as
# "compressed oops", in the JVM and make less effective use of the heap.
#
# Tip: Setting the initial and the max heap size to the same value means the
# JVM will never need to change the heap size. Changing the heap size otherwise
# involves a full GC, which is desirable to avoid.
#
# Based on the above, the following memory settings are recommended:
dbms.memory.heap.initial_size=4600m
dbms.memory.heap.max_size=4600m
dbms.memory.pagecache.size=4g
```

### CODE REVIEW

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+import static org.neo4j.io.ByteUnit.mebiBytes;
+import static org.neo4j.io.ByteUnit.tebiBytes;
+import static org.neo4j.kernel.configuration.Settings.BYTES;
+import static org.neo4j.kernel.configuration.Settings.buildSetting;
+
+public class MemoryRecommendationsCommand implements AdminCommand
+{
+    // Fields: {System Memory in GiBs; OS memory reserve in GiBs; JVM Heap memory in GiBs}.
+    // And the page cache gets what's left, though always at least 100 MiB.
+    // Heap never goes beyond 32 GiBs, which is the upper limit for compressed oops..
```
Comment (spacecowboy) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: You fail to mention Lucene and Netty off-heap memory needs. How are those taken into account here?
And what do you mean "oops"?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+                .withArgument( new OptionalBooleanArg( ARG_PRINT_CONF, false,
+                        "Print what an updated neo4j.conf file would look like to standard out. Does not change " +
+                        "existing settings unless --overwrite is specified." ) )
+                .withArgument( new OptionalBooleanArg( ARG_UPDATE_CONF, false,
+                        "Update the neo4j.conf file, or the given --conf-file, in-place by adding the recommended " +
+                        "memory settings, unless they are already specified. " +
+                        "Specify --overwrite to also update existing settings." ) )
+                .withArgument( new OptionalBooleanArg( ARG_OVERWRITE, false,
+                        "Causes --print-conf and --update-conf to overwrite existing settings, if set." ) )
+                .withArgument( new OptionalNamedArg( ARG_CONF_FILE, "conf" + File.separator + "neo4j.conf", "",
```
Comment (spacecowboy) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: This default path is incorrect for everything except the tarball. I think it would be easier for everybody if it just didn't have a default value.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+
+    public static Arguments buildArgs()
+    {
+        return new Arguments()
+                .withArgument( new OptionalNamedArg( ARG_MEMORY, "8g", OsBeanUtil.getTotalPhysicalMemory() + "",
+                        "Recommend memory settings with respect to the given amount of memory, " +
+                        "instead of the total memory of the system running the command." ) )
+                .withArgument( new OptionalBooleanArg( ARG_PRINT_CONF, false,
+                        "Print what an updated neo4j.conf file would look like to standard out. Does not change " +
+                        "existing settings unless --overwrite is specified." ) )
```
Comment (spacecowboy) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: I wonder if you should refer the the variable name here instead of hardcoding `--overwrite`

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 2 above)
Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: One reason for me saying so is that on Linux systems the user running the command will typically not have the same environment set as the Neo4j service does. So there will be no way to figure out where the config is.

A user will be invoking it like this

    neo4j-admin memrec .... /etc/neo4j/neo4j.conf

that command will fail because the user doesn't have write access to `/etc`. So second time around the user sudos:

    sudo neo4j-admin memrec .... /etc/neo4j/neo4j.conf

In neither case does the user have `NEO4J_CONF` set in their environment because that is set in the neo4j system service. So there is no way for the tool to figure out where to put the config in any reasonable automatic sense.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+            // The existing setting is enabled, but we are not allowed to overwrite it, so we return the line
+            // unmodified.
+            return line;
+        }
+        // We also need to make sure that what we have found, is not just a mention of the setting in some comment.
+        // We recognise commented out settings (as opposed to mentions of the setting in a comment) by the equals sign
+        // that follows the setting name, and the absence of any prose text before the setting name.
+        int endOfSettingIndex = settingIndex + name.length();
+        int equalsSign = line.indexOf( '=', endOfSettingIndex );
+        if ( equalsSign == -1 || containsProse( line, endOfSettingIndex, equalsSign ) ||
```
Comment (spacecowboy) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: Why do you differentiate between different kinds of comments?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        int settingIndex = line.indexOf( name );
+        if ( settingIndex == -1 )
+        {
+            // The setting we're looking for is not present in this line, so we just return it unmodified.
+            return line;
+        }
+        // Otherwise, we need to modify this line, such that it reflects the new value.
+        // Except, if 'overwrite' is false, then we don't modify the line if the setting is not commented out.
+        int firstComment = line.indexOf( '#' );
+        if ( !overwrite && (settingIndex < firstComment || firstComment == -1 || alreadyDefined.test( name )) )
```
Comment (spacecowboy) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: I'm confused about `alreadyDefined` in this statement. Why is it needed? And if it is needed, then why is it `OR`d with `!overwrite`?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandProvider.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                "can be expected to be dedicated to Neo4j.%n" +
+                "%n" +
+                "By default, the command will print its recommendations to the console in a human readable form. " +
+                "The command can alternatively print what the neo4j.conf file would look like with the new settings " +
+                "by specifying --print-conf, and those updated settings can be written back to the conf file " +
+                "(updated in-place) by specifying --update-conf. Any existing settings already enabled in the file " +
+                "will not be updated unless --overwrite is specified.%n" +
+                "%n" +
+                "If --print-conf or --update-conf is specified, then the command will by default look for, and work " +
+                "on, the standard neo4j.conf file. A different file can be specified with the --conf-file argument." );
```
Comment (spacecowboy) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandProvider.java]: See objection to "standard neo4j.conf file" above

File: community/dbms/src/test/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+                "# A comment.",
+                initialHeapSize.name() + "=" + heap + " # commented out setting",
+                maxHeapSize.name() + "=" + heap + " # commented out setting without value",
+                pagecache_memory.name() + "=" + pagecache
+        ) );
+        verify( outsideWorld ).stdErrLine( "File updated: " + confFile );
+    }
+
+    @Test
+    public void mustNotDefineSettingsMoreThanOnce() throws Exception
```
Comment (spacecowboy) [inline community/dbms/src/test/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandTest.java]: Plus 19 on this test 👍

File: community/dbms/src/test/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+        assertThat( recommendHeapMemory( gibiBytes( 1920 ) ), between( gibiBytes( 31 ), gibiBytes( 33 ) ) );
+    }
+
+    @Test
+    public void mustRecommendPageCacheMemory()
+    {
+        assertThat( recommendPageCacheMemory( mebiBytes( 100 ) ), between( mebiBytes( 95 ), mebiBytes( 105 ) ) );
+        assertThat( recommendPageCacheMemory( gibiBytes( 1 ) ), between( mebiBytes( 95 ), mebiBytes( 105 ) ) );
+        assertThat( recommendPageCacheMemory( gibiBytes( 3 ) ), between( mebiBytes( 470 ), mebiBytes( 530 ) ) );
+        assertThat( recommendPageCacheMemory( gibiBytes( 6 ) ), between( mebiBytes( 980 ), mebiBytes( 1048 ) ) );
```
Comment (spacecowboy) [inline community/dbms/src/test/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandTest.java:83]: How did you come up with the particular numbers? (I ask because no 2 people seem to recommend the same thing)

**Review by spacecowboy — CHANGES_REQUESTED** [review]
Excellent amount of tests.

How does it actually look when you run it? Could you copy paste the output of 

`neo4j-admin help`
`neo4j-admin help memrec`
`neo4j-admin memrec`

**Comment by spacecowboy** [comment]
I've changed the labels on the PR so it ends up in the correct place in the change log

File: community/dbms/src/test/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandTest.java (hunk 8 above)
Comment (chrisvest) [reply community/dbms/src/test/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommandTest.java:83]: I meditated on it for a bit, and then wrote down what felt reasonable.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 4 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: In `neo4j.conf` we have a passage that look like this:

```# The legacy policy is a special policy which is not defined in
# the policy configuration section, but rather derives from
# dbms.directories.certificates and associated files
# (by default: neo4j.key and neo4j.cert). Its use will be deprecated.
```

If a similar passage gets introduced by us or by an admin, for the settings that we are interested in here, then we should avoid clobbering it.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 1 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: Lucene and Netty do their own native memory stuff, so since that is neither heap nor pagecache, I account them in the OS bucket.

"oops" is "ordinary object pointer", and "compressed oops" is a JVM feature where object references are represented with 32 bits instead of 64 bits, when the heap is bellow a certain threshold.

Let me see if I can add some references for that.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 4 above)
Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: well sure, but we should just avoid modifying comments period. So I don't see the reason why we need to differentiate between different kinds of comments. Just ignore all lines where first non-whitespace character is `#`.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 1 above)
Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: OK. Then I just think you should mention that in the comment or somewhere else reasonable. It's one of the biggest surprises about Neo4js memory management and entirely non-obvious for 99% of users/devs.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+                        "memory settings, unless they are already specified. " +
+                        "Specify --overwrite to also update existing settings." ) )
+                .withArgument( new OptionalBooleanArg( ARG_OVERWRITE, false,
+                        "Causes --print-conf and --update-conf to overwrite existing settings, if set." ) )
+                .withArgument( new OptionalNamedArg( ARG_CONF_FILE, "conf" + File.separator + "neo4j.conf", "",
+                        "An alternative path to the configuration file, for use with --print-conf and --update-conf. " +
+                        "If left unspecified, the default neo4j.conf file location will be used." ) );
+    }
+
+    static String bytesToString( long bytes )
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: we already have an almost identical function in ByteUnit, any chances of using that instead of introducing new?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+    @Override
+    public void execute( String[] args ) throws IncorrectUsage, CommandFailed
+    {
+        Arguments arguments = buildArgs().parse( args );
+
+        Path confFile = arguments.getOptionalPath( ARG_CONF_FILE )
+                                 .orElse( configDir.resolve( Config.DEFAULT_CONFIG_FILE_NAME ) );
+        String memoryArgument = arguments.get( ARG_MEMORY );
+        long memory = buildSetting( ARG_MEMORY, BYTES, null ).build().apply( arguments::get );
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: we have a `buildSetting` function that does not require ugly null parameter

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+        {
+            out.accept( setting + "=" + value );
+        }
+    }
+
+    private void printRecommendation( String memoryArgument, String heap, String pagecache )
+    {
+        outsideWorld.stdOutLine(
+                "Based on the available memory, " + memoryArgument + ", and assuming the system is" +
+                " dedicated to running Neo4j,\nthe following memory settings are recommended:" );
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: please use `System.lineSeparator()` instead of `\n` if you want to add line separator here

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+import static org.neo4j.io.ByteUnit.mebiBytes;
+import static org.neo4j.io.ByteUnit.tebiBytes;
+import static org.neo4j.kernel.configuration.Settings.BYTES;
+import static org.neo4j.kernel.configuration.Settings.buildSetting;
+
+public class MemoryRecommendationsCommand implements AdminCommand
+{
+    // Fields: {System Memory in GiBs; OS memory reserve in GiBs; JVM Heap memory in GiBs}.
+    // And the page cache gets what's left, though always at least 100 MiB.
+    // Heap never goes beyond 32 GiBs, which is the upper limit for compressed oops..
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: We talked already earlier today about heaps that are bigger than 32g, i think we need to mention somewhere that our recommendations are for heaps lover than 32g. 
Also based on previously observed db's config values would be good to validate user defined 
values for common misconfigurations (like 33g heaps, for example).

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 1 above)
Comment (MishaDemianenko) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: I feel the pain here, do you think it will gonna improve the situation if we will gonna mention that in comments for memory related settings in default config template file?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+            printRecommendation( memoryArgument, heap, pagecache );
+        }
+
+        if ( updateConf )
+        {
+            ArrayList<String> updatedLines = new ArrayList<>();
+            outputUpdatedConf( confFile, heap, pagecache, overwrite, updatedLines::add );
+            try
+            {
+                Files.write( confFile, updatedLines, StandardCharsets.UTF_8 );
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: shouldn't we keep a backup of the original config file first and not simply override?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+            defineIfMissing( out, redefined, maxHeapSize.name(), heap );
+            defineIfMissing( out, redefined, pagecache_memory.name(), pagecache );
+        }
+        catch ( Exception e ) // Also catch runtime exceptions due to use of Streams.
+        {
+            throw new CommandFailed( "Failed to read configuration file: " + confFile, e );
+        }
+    }
+
+    private String tryUpdateSetting( String line, Predicate<String> alreadyDefined, Set<String> redefined,
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: this method and related trying to do a bunch of work by parsing and trying to understand user config file, in the same time it will fail to do so in some of the cases that we support during normal config loading since the method does not cover all possible format possibilities.
For example, multi-line lines or `:` as a separator just to name a few, so maybe instead of doing this attempt we will perform additional validation of already configured values or generate a suggested property values, so users can update config file if they think its relevant for them.
What do you guys think?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+        return brackets.recommend( Bracket::osMemory );
+    }
+
+    static long recommendHeapMemory( long totalMemoryBytes )
+    {
+        Brackets brackets = findMemoryBrackets( totalMemoryBytes );
+        return brackets.recommend( Bracket::heapMemory );
+    }
+
+    static long recommendPageCacheMemory( long totalMemoryBytes )
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java:83]: We have a heuristic currently that tries to determine correct page cache size in case if that was not configured explicitly. We need to merge those two things and have only one "recommender"

**Review by MishaDemianenko — COMMENTED** [review]
Have some of the comments, sorry if those were discussed already

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 1 above)
Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: We should mention it everywhere we talk about different buckets of memory

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 14 above)
Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: Personally I'm all for it. I think it would be sensible to do 

```
sudo neo4j-admin memrec >> /etc/neo4j/neo4j.conf
```

but I know support will really like an in-place update feature.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 1 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: To be clear, the "it" we want to mention is the fact that Lucene and Netty have their own native memory usage, right?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 9 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: It uses the wrong unit names. I would have merged them had this been in 3.4.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 11 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: Or it should be two separate calls to `stdOutLine`.

Comment (MishaDemianenko) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: absolutely, any of those that till make new line correct on different platforms

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 1 above)
Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: yes.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 13 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: I think we should rely on pipes on the command line instead of touching files at all.

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 4 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: And then instead just append the missing settings to the end of the file?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 9 above)
Comment (MishaDemianenko) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: maybe we can have a different variant of that function there something like `bytesToString( long bytes, Function<Unit,String> unitNameMapper) `? (also its used on in one internal place in 3.3 so I think we more than safe to update it in one way or another in 3.3 as well)

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 13 above)
Comment (MishaDemianenko) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: Fantastic, all my files related comments closed in this case then 🥋

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 14 above)
Comment (chrisvest) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: Taking your example literally, I think we can all find a solution that can satisfy us all:

```
sudo neo4j-admin memrec >> /etc/neo4j/neo4j.conf
```

All this needs to do, really, is just produce a _section_ of configuration file text, that can be appended to the end of an existing (or not) configuration file. There is no need parse any existing file. We just have to produce output in a format that works in a configuration file.

Comment (spacecowboy) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: my original example was incorrect. I naturally meant

    neo4j-admin memrec | sudo tee -a /etc/neo4j/neo4j.conf

But yes. The way the config *should* work is that the last defined value is used in case of duplicate values, so there should not be any need to change anything in the file.

That is of course how it *should* be. I know Java does that, I am less certain about our wrapper scripts parsing of the config. That's a separate issue though.

**Comment by chrisvest** [comment]
@spacecowboy @MishaDemianenko I believe your comments have been addressed. Please take another look.

**Review by spacecowboy — APPROVED** [review]
I like the output a lot! 👍 

```
$ bin/neo4j-admin memrec --memory=5G
# Memory settings recommendation from neo4j-admin memrec:
#
# Assuming the system is dedicated to running Neo4j and has 5G of memory,
# we recommend a heap size of around 2600m, and a page cache of around 768m,
# and that about 1800m is left for the operating system, and the native memory
# needed by Lucene and Netty.
#
# Tip: If the indexing storage use is high, e.g. there are many indexes or most
# data indexed, then it might advantageous to leave more memory for the
# operating system.
#
# Tip: The more concurrent transactions your workload has and the more updates
# they do, the more heap memory you will need. However, don't allocate more
# than 31g of heap, since this will disable pointer compression, also known as
# "compressed oops", in the JVM and make less effective use of the heap.
#
# Tip: Setting the initial and the max heap size to the same value means the
# JVM will never need to change the heap size. Changing the heap size otherwise
# involves a full GC, which is desirable to avoid.
#
# Based on the above, the following memory settings are recommended:
dbms.memory.heap.initial_size=2600m
dbms.memory.heap.max_size=2600m
dbms.memory.pagecache.size=768m
```

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+import static org.neo4j.kernel.configuration.Settings.BYTES;
+import static org.neo4j.kernel.configuration.Settings.buildSetting;
+
+public class MemoryRecommendationsCommand implements AdminCommand
+{
+    // Fields: {System Memory in GiBs; OS memory reserve in GiBs; JVM Heap memory in GiBs}.
+    // And the page cache gets what's left, though always at least 100 MiB.
+    // Heap never goes beyond 31 GiBs, which is the upper limit for compressed oops.
+    // More details on heap sizing wrt. compressed oops:
+    // https://www.elastic.co/guide/en/elasticsearch/guide/current/heap-sizing.html
```
Comment (MishaDemianenko) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java]: no 🙈 , let's talk with docs team about our own page about that kind of problems instead?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java (hunk 15 above)
Comment (MishaDemianenko) [reply community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java:83]: As I can see one part of the recommendation is gone, but heuristic that calculates size still there. Shouldn't our recommended value and heuristic evaluated be the same and use the same code for evaluation?

File: community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+import static org.neo4j.io.ByteUnit.ONE_MEBI_BYTE;
+import static org.neo4j.io.ByteUnit.gibiBytes;
+import static org.neo4j.io.ByteUnit.mebiBytes;
+import static org.neo4j.io.ByteUnit.tebiBytes;
+import static org.neo4j.kernel.configuration.Settings.BYTES;
+import static org.neo4j.kernel.configuration.Settings.buildSetting;
+
+public class MemoryRecommendationsCommand implements AdminCommand
+{
+    // Fields: {System Memory in GiBs; OS memory reserve in GiBs; JVM Heap memory in GiBs}.
```
Comment (jakewins) [inline community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java:47]: Note that Netty allocates off-heap RAM as well:

- there is no actual upper bound, it will allocate until malloc fails
- it is lazy, it allocates 16MiB chunks in "arenas:, by default there are `numCpu * 2` arenas, all with one chunk each

TL;DR: Assume Netty allocates 16 * numCPU * 2 MiB at bootstrap, and then unbound additional amounts until we modify the Causal Cluster and Bolt input code to use fixed size I/O buffers.

See https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java
